charmhelpers/contrib/openstack/deferred_events.py

```diff
@@ -0,0 +1,291 @@
+
+
+def get_deferred_events():
+    """Return a list of deferred events requested by the charm and packages.
+
+    :returns: List of deferred events
+    :rtype: List[ServiceEvent]
+    """
+    return DeferredCharmServiceEvents().events + DeferredPackageServiceEvents().events
+
+
+def get_deferred_restarts():
+    """List of deferred restart events requested by the charm and packages.
+
+    :returns: List of deferred restarts
+    :rtype: List[ServiceEvent]
+    """
+    return [e for e in get_deferred_events() if e.action == 'restart']
+
+
+def clear_deferred_events(services, action):
+    """Clear deferred events of type `action` targetted at `services`.
+
+    :param services: Services with deferred actions to clear.
+    :type services: List[str]
+    :param action: Action to clear
+    :type action: str
+    """
+    DeferredCharmServiceEvents().clear_deferred_events(services, action)
+    DeferredPackageServiceEvents().clear_deferred_events(services, action)
+
+
+def clear_deferred_restarts(services):
+    """Clear deferred restart events targetted at `services`.
+
+    :param services: Services with deferred actions to clear.
+    :type services: List[str]
+    """
+    clear_deferred_events(services, 'restart')
+
+
+def process_svc_restart(service):
+    """Respond to a service restart having occured.
+
+    :param service: Services that the action was performed against.
+    :type service: str
+    """
+    clear_deferred_restarts([service])
+
+
+def is_restart_permitted():
+    """Check whether restarts are permitted.
+
+    :returns: Whether restarts are permitted
+    :rtype: bool
+    """
+    if hookenv.config('enable-auto-restarts') is None:
+        return True
+    return hookenv.config('enable-auto-restarts')
+
+
+def defer_restart_on_changed(service, changed_files):
+    """Check if restarts are permitted, if they are not defer them.
+
+    :param service: Service to be restarted
+    :type service: str
+    :param changed_files: Files that have changed to trigger restarts.
+    :type changed_files: List[str]
+    :returns: Whether restarts are permitted
+    :rtype: bool
+    """
+    permitted = is_restart_permitted()
+    if not permitted:
+        charm_events = DeferredCharmServiceEvents()
+        charm_events.add_event(
+            service,
+            event_reason='File(s) changed: {}'.format(
+                ', '.join(changed_files)),
+            action='restart')
+    return permitted
+
+
+def deferrable_svc_restart(svc, reason=None):
+    """Restarts service if permitted, if not defer it.
+
+    :returns: Whether restarts are permitted
+    :rtype: bool
+    """
+    if is_restart_permitted():
+        host.service_restart(svc)
+    else:
+        charm_events = DeferredCharmServiceEvents()
+        charm_events.add_event(
+            svc,
+            event_reason=reason,
+            action='restart')
+
+
+def configure_deferred_restarts(services):
```

Comment:
Is this specific to package restarts? The function name may need to reflect that.

Reply:
Again I was thinking of future proofing. The charms call this method with the idea that it will ensure that all setup for deferrable restarts is done. It's just that there is no setup needed for charm deferred events.